### DISCUSSION TITLE
How can I control the access to the data?

### DISCUSSION THREAD

**Discussion by Spottel** [MichalLytek/typegraphql-prisma#72] [Q&A]
I need your help.
I use typegraphql-prisma with a mysql database.
I have an userAuthChecker Function to control which user/user group can use which resolver with an x-access-token.
And that works. 
But now I have one problem. I have a table "projects". This table contains the columns "id;user_id;name...."
I can view the data with findManyProjects. But every logged user can view all data, not only the own data.
How can I control that User 1 only can view User 1 Projects Data?
At the moment it is possible with a valid x-access-token to make a request to the server and view all data.

**Comment by MichalLytek** [comment]
If you have custom rules and business logic, create custom resolvers for that 😉 
```ts
import { Client, FindManyClientArgs } from "@generated/type-graphql";

@Resolver(of => Client)
class ClientResolver {
  // use generated model type
  @Query(returns => [Client])
  async customFindClientsWithArgs(
    // use generated args type
    @Args() args: FindManyClientArgs,
    @Ctx() { prisma }: Context,
  ): Promise<Prisma.User[]> {
    // put your logic here
    // like adding some where conditions to args based on `context.userId`

    // call prisma normally
    return prisma.user.findMany(args);
  }
}
```

**Comment by repomaa** [comment]
I think this is a pretty common use case. In my case I quickly realized, that i can't use almost any of the (otherwise really cool) generated query resolvers. It's a lot of boilerplate to write the `where`, and `sort` and other functionality by hand. This is why I'm proposing the following (without knowing if this would actually be possible or not):

Add a new decorator `EnhanceQuery` or something, which receives the `XYZFindManyArgs` (or `XYZFindFirstArgs` etc.) object as generated from the args and allows merging it with additional custom clauses. Then we could do something like:

``` ts
const resolversEnhanceMap: ResolversEnhanceMap = {
  Project: {
    projects: [
      EnhanceQuery(
        (
          { context: { user: { id: userId },
          args: ProjectFindManyArgs
        ) => ({
          ...args,
          where: { ...args.where, userId }
        })
      )
    ]
  }
};
```

**Reply by MichalLytek** [reply]
> I think this is a pretty common use case. In my case I quickly realized, that i can't use almost any of the (otherwise really cool) generated query resolvers.

`typegraphql-prisma` is not a universal tool to expose your db via GraphQL, like Hasura and others.
It's primarily a generator of TypeGraphQL classes, so you don't need to write model's object types, inputs and args.
Generated CRUD resolvers are intended mainly for rapid prototyping, not production usage nor with ability to add custom logics.

So  proposing stuff like `EnhanceQuery` is strongly against the design goals. If it would be a universal approach, it wouldn't generate TypeGraphQL classes at all, just build a schema, have a special language for declaring exposed fields/operations, as well as hook/plugins based system for attaching logic before-after calling resolver/field.

So if you need something more sophisticated like validation or auth, you should just create (copy) the generated resolver and add your business logic inside the custom resolver method.

**Reply by repomaa** [reply]
I respect that. In hindsight I think it's better that way, keeps the code simpler. I do think, however, that the docs suggest you _should_ be using the generated crud resolvers. Cherry-picking and decorating is under "Advanced", after all. I don't have a problem with that but it think it would be helpful for people stumbling across this project, to see the docs emphasize the usage of the generated object and input types for their custom api, like they're bound to do in production anyway.

**Comment by repomaa** [comment]
@Spottel for now, I think this is a nice pattern to use for modifying args to resolvers:

```ts
@Resolver()
export default class Tokens {
  private original: FindManyTokenResolver // imported from generated code

  constructor() {
    this.original = new FindManyTokenResolver()
  }

  @Query((returns) => [Token])
  @Authorized()
  async tokens(
    @Ctx() ctx: Context,
    @CurrentUser() { id: userId }: User,
    @Args() originalArgs: FindManyTokenArgs // imported from generated code
  ) {
    const args: FindManyTokenArgs = {
      ...originalArgs,
      where: { ...originalArgs.where, userId: { equals: userId } },
    }

    return this.original.tokens(ctx, args)
  }
}
```

**Reply by repomaa** [reply]
I couldn't extend the original resolver because I added the current user to the arguments with my `@CurrentUser()` decorator. If you pull the user from ctx every time, that would work too. I just added the decorator for convenience so I don't need to cast it in the function (i'm using typescript with `strict` and user can be undefined)

**Reply by MichalLytek** [reply]
What's the purpose of reusing that resolver if it only passes params to `prisma` inside?
```ts
@Resolver()
export default class Tokens {
  @Query((returns) => [Token])
  @Authorized()
  async tokens(
    @Ctx() ctx: Context,
    @CurrentUser() { id: userId }: User,
    @Args() originalArgs: FindManyTokenArgs // imported from generated code
  ) {
    const args: FindManyTokenArgs = {
      ...originalArgs,
      where: { ...originalArgs.where, userId: { equals: userId } },
    }
    return ctx.prisma.token.findMany(args);
  }
}
```

**Reply by repomaa** [reply]
Hm I guess nothing. :) But like this, i don't need to know what the generated resolver does

**Reply by MichalLytek** [reply]
It's a TypeGraphQL building block, not a black box magic, at least it's not intended for using that way 😅

**Reply by repomaa** [reply]
Sure, I didn't mean it that way, but It's different for you having written the generator. It's usually safer to call the original method of some library you don't have the time to look into the source code. This way my code will probably work even if you for whatever reason decide to do more than just `ctx.prisma.token.findMany(args)` in there. e.g. i didn't know how field renaming is handled, so in case it's done by some hook in a generated resolver function, i'd rather just call it, to be safe.

**Reply by MichalLytek** [reply]
Thanks for the feedback. I will spend next iteration on better docs and examples, so users won't have to invent their own approach for common cases 😉

**Reply by dbismut** [reply]
I'm a bit late to the party, but just discovered typegraphql and immediately stumbled upon this issue. It looks like you can actually mutate `args` in custom a decorator via `createMethodDecorator` applied with `applyResolversEnhanceMap`. 

```ts
function FilterOrders() {
  return createMethodDecorator<{ user: Express.User }>(async ({ args, context }, next) => {
    const { user } = context

    args.where = { ...args.where, userId: { equals: user.id } }
    
    return next()
  })
}
```
